////////////////////////////////////////////////////////////////////////////////
// Copyright (c) 2014-2019, Lawrence Livermore National Security, LLC.
// Produced at the Lawrence Livermore National Laboratory.
// Written by the LBANN Research Team (B. Van Essen, et al.) listed in
// the CONTRIBUTORS file. <lbann-dev@llnl.gov>
//
// LLNL-CODE-697807.
// All rights reserved.
//
// This file is part of LBANN: Livermore Big Artificial Neural Network
// Toolkit. For details, see http://software.llnl.gov/LBANN or
// https://github.com/LLNL/LBANN.
//
// Licensed under the Apache License, Version 2.0 (the "Licensee"); you
// may not use this file except in compliance with the License.  You may
// obtain a copy of the License at:
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
// implied. See the License for the specific language governing
// permissions and limitations under the license.
////////////////////////////////////////////////////////////////////////////////

syntax = "proto3";

package lbann_data;

import "callbacks.proto";
import "layers.proto";
import "metrics.proto";
import "objective_functions.proto";
import "weights.proto";

message Model {
  message Summarizer {
    string dir = 1;
  }

  string type = 1;
  string name = 3;
  ObjectiveFunction objective_function = 2;
  repeated Metric metric = 5;
  string data_layout = 6;

  int64 num_epochs = 4;
  int64 super_steps = 121; //multiple steps/epochs currently use in GAN
  int64 num_batches = 122; //multiple batches/sub epoch
  int64 evaluation_frequency = 54;
  bool  serialize_io = 101;

  bool disable_cuda = 8;

  repeated Layer layer = 10;

  repeated Weights weights = 11;

  repeated Callback callback = 20;

  Summarizer summarizer = 32;
}
